data_filter_mongodb/internal/opa/opa.go

```diff
@@ -0,0 +1,203 @@
+		rego.ParsedInput(inputTerm.Value),
+		rego.Unknowns(unknowns),
+	)
+
+	pq, err := r.Partial(ctx)
+	if err != nil {
+		c.logger.Error("Failed to perform opa partial eval", zap.Error(err))
+		return mongo.Result{}, err
+	}
+
+	if len(pq.Queries) == 0 {
+		// always deny
+		return mongo.Result{Defined: false}, nil
+	}
+
+	for query := range pq.Queries {
+		if len(pq.Queries[query]) == 0 {
+			// always allow
+			return mongo.Result{Defined: true}, nil
+		}
+	}
+
+	return c.processQuery(pq)
+}
+
+func (c *opaClient) processQuery(pq *rego.PartialQueries) (mongo.Result, error) {
+	var queries []mongo.Queries
+	c.logger.Info("opa-query: ", zap.Any("queries", fmt.Sprintf("%s", pq.Queries)))
+	for i := range pq.Queries {
+		pipeline := &[]bson.M{}
+		for _, expr := range pq.Queries[i] {
+			if !expr.IsCall() {
+				continue
+			}
+
+			if len(expr.Operands()) != 2 {
+				return mongo.Result{}, fmt.Errorf("invalid expression: too many arguments")
+			}
+
+			var value interface{}
+			var processedTerm []string
+			var err error
+			for _, term := range expr.Operands() {
+				if ast.IsConstant(term.Value) {
+					value, err = ast.JSON(term.Value)
+					if err != nil {
+						return mongo.Result{}, fmt.Errorf("error converting term to JSON: %v", err)
+					}
+				} else {
+					processedTerm = processTerm(term.String())
+				}
+			}
+
+			if processedTerm == nil {
+				return mongo.Result{}, nil
+			}
+
+			if isEqualityOperator(expr.Operator().String()) {
+				mongo.HandleEquals(pipeline, processedTerm[1], value)
+			} else if isRangeOperator(expr.Operator().String()) {
+				if expr.Operator().String() == "lt" {
+					mongo.HandleLessThan(pipeline, processedTerm[1], value)
+				} else if expr.Operator().String() == "gt" {
+					mongo.HandleGreaterThan(pipeline, processedTerm[1], value)
+				} else if expr.Operator().String() == "lte" {
+					mongo.HandleLessThanEquals(pipeline, processedTerm[1], value)
+				} else if expr.Operator().String() == "gte" {
+					mongo.HandleGreaterThanEquals(pipeline, processedTerm[1], value)
+				}
+			} else if expr.Operator().String() == "neq" {
+				mongo.HandleNotEquals(pipeline, processedTerm[1], value)
+			} else {
+				return mongo.Result{}, fmt.Errorf("invalid expression: operator not supported: %v", expr.Operator().String())
+			}
+		}
+		k1 := mongo.Queries{Pipeline: bson.M{"$and": *pipeline}}
+		queries = append(queries, k1)
+	}
+	result := c.mongoClient.QueryMongo(queries)
+
+	return result, nil
+}
+
+func processTerm(query string) []string {
+	splitQ := strings.Split(query, ".")
+	var result []string
+	for _, term := range splitQ {
+		result = append(result, removeOpenBrace(term))
+	}
+
+	if result == nil {
+		return nil
+	}
+
+	indexName := result[1]
+	fieldName := result[2]
+	if len(result) > 2 {
+		fieldName = strings.Join(result[2:], ".")
+	}
+
+	return []string{indexName, fieldName}
+}
+
+func removeOpenBrace(input string) string {
+	return strings.Split(input, "[")[0]
```

Comment:
Why is this function needed ?

Reply:
This function is used in processTerm call to remove brackets to process terms. I.e.
```
func processTerm(query string) []string {
	splitQ := strings.Split(query, ".")
	var result []string
	for _, term := range splitQ {
		result = append(result, removeOpenBrace(term))
	}

```